src/boot/efi/chid.c

```diff
@@ -0,0 +1,385 @@
+                break;
+        case 7:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->product_sku);
+                break;
+        case 8:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                7,
+                                info->manufacturer,
+                                L"&",
+                                info->product_name,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 9:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->product_name);
+                break;
+        case 10:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                7,
+                                info->manufacturer,
+                                L"&",
+                                info->family,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
+                break;
+        case 11:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                3,
+                                info->manufacturer,
+                                L"&",
+                                info->family);
+                break;
+        case 13:
+                status = hash_strings_sha1(
+                                &hash,
+                                (uint8_t *) &namespace,
+                                sizeof(namespace),
+                                5,
+                                info->manufacturer,
+                                L"&",
+                                info->baseboard_manufacturer,
+                                L"&",
+                                info->baseboard_product);
```

Comment:
so, i find the implementation with structs and fields, and switch statements and varargs a bit too complex. what about changing `info` to just be a string array, indexed by an enum:

```
typedef ChidSmbiosFields {
       CHID_SMBIOS_MANUFACTURER,
       CHID_SMBIOS_BASEBOARD_MANUFACTURER,
       …
} ChidSmbiosFields;
```

And then you could write `info[CHID_SMBIOS_MANUFACTURER]` where you write `info->manufacturer` now.

Now, that doesn't seem like much of an improvement, but once you did that, you can replace this switch/case statement by this:

```c
static const uint32_t chid_smbios_table[] = {
        [3] = (UINT32_C(1) << CHID_SMBIOS_MANUFACTURER) |
              (UINT32_C(1) << CHID_FIELD) |
              …,
        [4] = …
}
```

you get the idea.

and then just iterate through the bits and include in the hash whereever the bit is set.